BUG: fix signed zero behavior in npy_divmod

tlatorre-uchicago opened the pull request:
First attempt to fix #12841. I've only minimally tested this so far, so it shouldn't be merged as is yet. I plan to add a few test functions and to look and see if there are any other places in the code where `copysign()` should be used in place of a simple check against zero.

tlatorre-uchicago commented:
Ok, I've added a new test for this and it should be good to go. @charris since it looks like you were the one who originally updated divmod to have the same behavior as python's divmod could you take a look at this and make sure it looks OK?

eric-wieser commented on numpy/core/tests/test_umath.py:
```diff
@@ -273,6 +273,12 @@ def test_floor_division_complex(self):
+        # negative zero by one.
+        x = np.zeros(10)
+        assert_((~np.signbit(x//1)).all())
+        assert_(np.signbit((-x)//1).all())
```
nit: probably better as
```python
assert_equal(np.signbit(x//1), 0)
assert_equal(np.signbit((-x)//1), 1)
```

ahaldane commented:
Looks correct to me, the code now does exactly what python3's `float_divmod` does. Great!

A few more steps before merging:

 1. I think this is worth a compatibility note. In the file `doc/release/1.17.0-notes.rst`, could you add a note in the "compatibility section" mentioning that the sign of zero returned by floor_divide and divmod will change?
 2. Also, in one of the commit messages, could you put the sentence "fixes #12841". That way github will auto-close that issue when this is merged.
 3.  Also, as far as I can see, #6317, for the sign of `np.remainder` is already fixed in master. If you add a test for the `remainder` zero-sign in this PR, then I think it would be fair to add a `fixes #6317` into the commit message here. In the compatibility note above you might also mention that remainder (already) also follows this zero-sign behavior.

I also think #7709 should not be closed with this PR, since it is mainly a debate about whether 1.0/0.0 should return `inf` or `nan` which does not appear to be resolved, though maybe it's too late to decide that now. So nothing to do about that here.

With those changes I think this would be good to merge.

tlatorre-uchicago commented:
1. Ok I will add something to the notes. Is it worth mentioning the fact that it used to be correct here somewhere too? I haven't been able to narrow down the exact commit because older versions won't build for me here, but I know on older python versions it doesn't have this behavior. I suspect it was this commit: https://github.com/numpy/numpy/commit/25c23f1d956104a072a95355ffaa7a38b53710b7.
2. Sure, will do.
3. Huh, that must have been a really recent change. I'm fine with just mentioning #12841.

Yea #7709 had a bunch of issues, one of which was this behavior, but there were other ones too.

ahaldane commented:
Agreed, I also came across the commit you found, that's probably when it changed.

Yeah, I think it's good to briefly mention that older numpy had the correct sign too, it helps justify the change.

charris commented:
@tlatorre-uchicago Could you add the release note? Probably under `Compatibility`.

tlatorre-uchicago commented:
Do I create a new file called 1.16.2-notes.rst in the doc directory? Also, can you verify which release changed the initial behavior?

tlatorre-uchicago commented:
I added an update to the 1.17.0 release notes. I assumed that the commit I mentioned earlier was the initial cause of the bug and so the bug was introduced in 1.12.0.

seberg commented on numpy/core/tests/test_umath.py:
```diff
@@ -273,6 +273,12 @@ def test_floor_division_complex(self):
+        # negative zero by one.
+        x = np.zeros(10)
+        assert_equal(np.signbit(x//1),0)
+        assert_equal(np.signbit((-x)//1),1)
```
Can you add PEP8 conform spaces after the `,`? Spaces around `//` are optional.

tlatorre-uchicago replied:
This has been updated in my latest push.

seberg commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -654,7 +654,7 @@ npy_divmod@c@(@type@ a, @type@ b, @type@ *modulus)
     else {
         /* if mod is zero ensure correct sign */
-        mod = (b > 0) ? 0.0@c@ : -0.0@c@;
+        mod = npy_copysign@c@(0, b);
```
I guess this effectively does not matter (since the result will always be NaN if b is 0). Would be nice to add a test for this path anyway (I assume it is missing right now).
I suppose this might be marginally slower, but I guess it should be very little and it is nicer to read/uniform.

tlatorre-uchicago replied:
Can you elaborate on what path you are referring to? The `b == 0` case is handled at the top of the function.

seberg replied:
Sorry, was just a silly comment. `b != 0` is guaranteed here I think, so the old comparison was OK.

tlatorre-uchicago replied:
Oh I see what you mean. Yeah, I think you're right. Do you think it's ok to keep this for consistency with the cpython version here https://github.com/python/cpython/blob/6d43f6f081023b680d9db4542d19b9e382149f0a/Objects/floatobject.c#L648 or should I revert it?

seberg replied:
Yes, I think it is OK. If it was much slower that would be an argument, but I cannot imagine it. I just stumbled on it, because I tried to make a test related to this change.

seberg commented on doc/release/1.17.0-notes.rst:
```diff
@@ -36,7 +36,20 @@ Casting from a different floating point precision to float16 used incorrect
+Signed zero when using divmod
+-----------------------------
 
+Starting in version 1.12.0, numpy incorrectly return a negatively signed zero
```
return -> returned. Maybe "``` ``divmod`` ``` and ``` ``floor_devide`` ```" (add the double ticks also, I think), since while this is clearer from implementation wise, `//` is much more often used (which is why you used it as an example after all).

tlatorre-uchicago replied:
Ok, this has been updated in my latest push.

seberg reviewed (approved):
Looks good to me, thanks @tlatorre-uchicago. If you do not have time to do another iteration, just say so, and may just fix it up a bit.

tlatorre-uchicago committed:
BUG: fix signed zero behavior in npy_divmod

Previously when doing floor division numpy would sometimes return an incorrect
signed zero. For example:

>>> np.zeros(10)//1
array([-0., -0., -0., -0., -0., -0., -0., -0., -0., -0.])

>>> np.remainder(-1.0,1.0)
-0.0

The reason for this is that whenever div or mod were zero the code was using
the following to pick the sign of zero:

floordiv = (a / b > 0) ? 0.0@c@ : -0.0@c@;

This commit updates these lines to instead use the copysign function which is
how cpython does floor division.

Fixes #12841.

charris commented:
Thanks @tlatorre-uchicago .

tlatorre-uchicago commented:
Awesome, thanks!
